topostats/io.py

```diff
@@ -986,6 +986,49 @@ def add_to_dict(self, image: np.ndarray, filename: str) -> None:
         }
 
 
+def _hdf5_add_known_datatype(
+    open_hdf5_file: h5py.File, group_path: str, item: list | str | int | float | np.ndarray | Path | dict, key: str
+) -> None:
+    """Add a known datatype to an open hdf5 file."""
+    # Lists need to be converted to numpy arrays
+    if isinstance(item, list):
+        item = np.array(item)
+    # Strings need to be encoded to bytes
+    elif isinstance(item, str):
+        open_hdf5_file[group_path + key] = item.encode("utf8")
+    # Integers, floats and numpy arrays can be added directly to the hdf5 file
+    # Ruff wants us to use the pipe operator here but it isn't supported by python 3.9
+    elif isinstance(item, (int, float, np.ndarray)):  # noqa: UP038
+        open_hdf5_file[group_path + key] = item
+    # Path objects need to be encoded to bytes
+    elif isinstance(item, Path):
+        open_hdf5_file[group_path + key] = str(item).encode("utf8")
+    # Dictionaries need to be recursively saved
+    elif isinstance(item, dict):  # a sub-dictionary, so we need to recurse
+        dict_to_hdf5(open_hdf5_file, group_path + key + "/", item)
+
+
+def dict_to_hdf5(open_hdf5_file: h5py.File, group_path: str, dictionary: dict) -> None:
+    """Recursively save a dictionary to an open hdf5 file."""
+    for key, item in dictionary.items():
+        LOGGER.info(f"Saving key: {key}")
+
+        if item is None:
+            LOGGER.warning(f"Item '{key}' is None. Skipping.")
+        # Make sure the key is a string
+        key = str(key)
+
+        # Check if the item is a known datatype
+        # Ruff wants us to use the pipe operator here but it isn't supported by python 3.9
```

Comment:
You can probably get round that and support it in Python 3.9 by...

```python
from __future__ import annotations
```

Reply:
`io.py` already seems to have that at the top of the file
```python
"""Functions for reading and writing data."""
from __future__ import annotations

import importlib.resources as pkg_resources
import io
import logging
.
.
.
```

Reply:
:facepalm: sorry for not checking, in which case we could...

1. Disable with `# noqa: <ruff_code>`
2. Make the minimum Python version 3.10

I'm somewhat inclined to go with 2. myself. I was working on Skan earlier this week and discovered NetworkX supports 3.10 and higher so things are slowly being phased out (our tests on Skan fail under 3.9 so need to investigate why before removing support in case its indicative of a problem that is being masked in newer versions).

Reply:
I'd be in support of making the minimum Python 3.10.